### ISSUE TITLE
promql: sort_by_label should fallback to a consistent ordering by metric, not by sample value

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#14289]
### Proposal

I got an idea while thinking about https://github.com/prometheus/prometheus/pull/12503 : 

Currently, if `sort_by_label` encounters the same label values to sort by, it falls back to sorting by sample value. However, by using `sort_by_label`, the user essentially tells us they aren't interested in the sample value, but in label values.

I guess it would be more useful to use the rest of the metric (label values, name) as the "tie breaker" rather than the sample value. It doesn't even have to be a particular order, as long as the tie breaker yields consistent results, i.e. we could just use the metric hash as the tiebreaker rather than the sample value.

In other words, the sorting would always be the same, even if sample values change.

**Comment by beorn7** [comment]
The function is still experimental, so we can change the behavior.

An additional bonus is that we don't have to look at the sample value at all anymore, which might allow some optimizations.

@juliusv @bboreham WDYT?

**Comment by roidelapluie** [comment]
I agree with your proposal

**Comment by juliusv** [comment]
@beorn7 I agree!

**Comment by bboreham** [comment]
I can't think how it can happen that two series have the same label value.
The PromQL engine checks between each operation and halts with an error if it finds this case.

However I agree it is a nice optimisation, to avoid doing work that cannot be useful.

**Comment by bboreham** [comment]
Oh, my mistake - it sorts by a subset of labels, which can have the same value.

**Comment by suntala** [comment]
May I please be assigned to this issue?
